### ISSUE TITLE
StrawberryShake v12 that says it supports the graphql-ws protocol defaults its subprotocol to the wrong spec

### ISSUE THREAD

**Issue by elemental1313** [ChilliCream/hotchocolate#5196]
### Is your feature request related to a problem?

Hello!  I saw that the StrawberryShake  GQL Client had documentation to support the graphql-ws protocol as located here: https://chillicream.com/docs/strawberryshake/networking#protocols 

But in the code I see that the default protocol that gets added is "graphql-ws" and not "graphql-transport-ws" (which the [official documentation here](https://github.com/enisdenjo/graphql-ws/blob/master/PROTOCOL.md#communication) states should be the correct subprotocol, I know, its weird that the project is called graphql-ws but the sub protocol they expect is graphql-transport-ws) which is what my graphql server im trying to connect to is using (not currently using hotchocolate) and the traffic manager that makes sure the protocols match up give me an error, since i cannot specify only "graphql-transport-ws" as the sole protocol. 

Here is the file in question that has that default set: https://github.com/ChilliCream/hotchocolate/blob/d412413d5f93f74e897c1253e4786f61fc59f773/src/StrawberryShake/Client/src/Transport.WebSockets/Protocols/GraphQLWebSocket/GraphQLWebSocketProtocolFactory.cs

### The solution you'd like

I would love the ability to be able to set my own default protocol or have the option to configure my StrawberryShake client to use graphql-transport-ws protocol for subscriptions.  I assume the change would be in this file: https://github.com/ChilliCream/hotchocolate/blob/d412413d5f93f74e897c1253e4786f61fc59f773/src/StrawberryShake/Client/src/Transport.WebSockets/Protocols/GraphQLWebSocket/GraphQLWebSocketProtocolFactory.cs

(I tried to make an edit myself to submit as a suggestion but it didnt work locally how i thought it would, so wondering if someone in the community can help me or maybe look into this to see if im correct or not) 

### Product

Strawberry Shake

**Comment by kstaas1** [comment]
Having same issue after tying out this package. So full disclosure I'm just another potential consumer and and you may need to take what I have to say with a grain of salt. But from my investigation there isn't so much a default protocol here as there is only one available protocol implemented for the client at the moment. You can certainly add additional protocols to the connection request to the server via calling ConfigureWebSocketClient, however it will do little good as the client will immediately close the connection since it does not have an implementation for the protocol.

If you were felling like getting it working yourself it looks like you would need a separate IProtocolFacotry for the protocol (graphql-transport-ws) as well as a proper implementation of the ISocketProtocol interface that the factory returns that appropriately handles the graphql-transport-ws protocol. Then you could register that factory (presumably via AddProtocol on she ServiceCollection) and have it available for use with socket communication. Sadly I as mentioned did not see such implementation when searching in the code which means you would have to implement yourself. 

TLDR; from what I can tell the client support for graphql-transport-ws protocol does not seem to be available (yet?). Perhaps the new protocol may be only supported in the server side code but at the moment not the client which is why its mentioned in the documentation?

**Comment by elemental1313** [comment]
Still looking for some assistance on this, perhaps in the future when I have time i could propose a solution but i just dont :(

**Comment by michaelstaib** [comment]
We will add new subscription protocols in 13 ... this is on our backlog.

**Comment by elemental1313** [comment]
Looking forward to it! Thanks!

**Comment by kmcclellan** [comment]
I'm confused about the current state of subscription protocols after reading the documentation (which promises to deliver SignalR and GRPC in version 12?)

Can we get an update from the maintainers here?

**Comment by kmcclellan** [comment]
If I'm looking in the right place, it appears that currently [Hot Chocolate also only supports the outdated protocol](https://github.com/ChilliCream/graphql-platform/blob/d412413d5f93f74e897c1253e4786f61fc59f773/src/HotChocolate/AspNetCore/src/AspNetCore/Subscriptions/WebSocketConnection.cs#L13)?

The reason I'm asking is that I need to know how GraphQL subscriptions handle unexpected connection interruptions. `graphql-transport-ws` includes `Ping` and `Pong` to support detecting these. Is there any documentation on this other protocol?

**Comment by elemental1313** [comment]
For that older stuff you could try this [link](https://github.com/enisdenjo/graphql-ws/blob/v0.0.2/PROTOCOL.md) which is just graphql-ws's older version of the protocol i linked above.  Doing a quick search, you are looking for the old protocol's "GQL_CONNECTION_KEEP_ALIVE".  I suggest searching the repo for the phrase "keep alive" and you will see some good tests and other files that should help you.  But in the mean time, this post and thread is about getting the graphql-transport-ws protocol type added to strawberryshake client for websockets 😄

**Comment by Dominig12** [comment]
> If I'm looking in the right place, it appears that currently [Hot Chocolate also only supports the outdated protocol](https://github.com/ChilliCream/graphql-platform/blob/d412413d5f93f74e897c1253e4786f61fc59f773/src/HotChocolate/AspNetCore/src/AspNetCore/Subscriptions/WebSocketConnection.cs#L13)?
> 
> The reason I'm asking is that I need to know how GraphQL subscriptions handle unexpected connection interruptions. `graphql-transport-ws` includes `Ping` and `Pong` to support detecting these. Is there any documentation on this other protocol?

What is provided here is the HotChocolate version when using ASP.NET, but in the implementation for the .NET console application (from ASP.NET) StrawberryShake and HotChocolate have different protocols ("graphql-ws" for the first and "graphql-transport-ws" for the second)
Because of this, the problem arises that I cannot connect the client from StrawberryShake and the server part from HotChocolate

**Comment by elemental1313** [comment]
> Because of this, the problem arises that I cannot connect the client from StrawberryShake and the server part from HotChocolate

Now that would make a nicer case i think for updating StrawberryShake to include this protocol. But i do believe that HotChocolate has its own client that you can use.  But the StrawberryShake client is VERY cool, basically automatically writing code for me to connect to my schema and allow me to make requests.  I do hope they add this to their backlog.  Again I am still very busy in my day job otherwise id be happy to try to help.  Until then, all I can do is hold hope in my heart 😄 ❤️

**Comment by michaelstaib** [comment]
Strawberry Shake currently supports graphql-sse and the subprotocol: graphql-ws 

```
    /// <summary>
    /// The sub-protocol name for the GraphQL over WebSocket Protocol from Apollo
    /// https://github.com/apollographql/subscriptions-transport-ws/blob/master/PROTOCOL.md
    /// see constants:
    /// https://github.com/apollographql/subscriptions-transport-ws/blob/master/src/protocol.ts
    /// </summary>
    public const string GraphQL_WS = "graphql-ws";
```

We will towards end of this year (probably autumn) shift resources to Strawberry Shake to create a new major version which will change the status quo on protocols.

**Comment by jimitndiaye** [comment]
> Strawberry Shake currently supports graphql-sse and the subprotocol: graphql-ws 
> 
> ```
>     /// <summary>
>     /// The sub-protocol name for the GraphQL over WebSocket Protocol from Apollo
>     /// https://github.com/apollographql/subscriptions-transport-ws/blob/master/PROTOCOL.md
>     /// see constants:
>     /// https://github.com/apollographql/subscriptions-transport-ws/blob/master/src/protocol.ts
>     /// </summary>
>     public const string GraphQL_WS = "graphql-ws";
> ```
> 
> We will towards end of this year (probably autumn) shift resources to Strawberry Shake to create a new major version which will change the status quo on protocols.


@michaelstaib 
It is mentioned here that strawberry shake supports graphql-sse but I haven't been able to get it to work. Setting the subscription transport to HTTP in the .graphqlrc doesn't work at runtime because while strawberry shake successfully connects to the server it doesn't actually react to individual messages so the client is left waiting forever. Switching back to websockets with the same code makes everything work normally. I can't find any examples of strawberry shake over SSE either. Ive posted about this on slack and github and haven't been able to get a response.
